File: LInear Data Structures/stackLinkedList.java
// stack as linked list built by user's input

import java.util.*;
import static java.lang.System.exit;

class stackLinkedlist { 
   
    // Define Node of LinkedList 
    private class Node { 
        int value; // node named as value 
        Node nodeLink; // Node link 
    } 

    // top of the stack 
    Node topNode; 
    // stack class Constructor 
    stackLinkedlist() { 
        this.topNode = null; 
    } 
   
    // push () method 
    public void push(int nodeToInsert) {
        // create a new node 
        Node tempNode = new Node();
   
        // assign value to node 
        tempNode.value = nodeToInsert; 
   
        // set top node of the stack to node link
        tempNode.nodeLink = topNode; 
   
        // update top to be the temporary node
        topNode = tempNode; 
    }   

      // isEmpty () method
    public boolean isEmpty() { 
        return topNode == null; 
    } 
 
    // peek () method 
    public int peek() { 
        // check if the stack is empty 
        // if the stack is not empty
        if (!isEmpty()) { 
            // return the value of the top node
            return topNode.value; 
        } 
        else { 
            System.out.println("Stack is empty!"); 
            return -1; 
        } 
    } 
   
    // pop () operation 
    public void pop() { 
        // check if stack is out of elements 
        if (topNode == null) { 
            System.out.print("\nStack Underflow!!"); 
            return; 
        } 
        // set top node to point to next node 
        topNode = (topNode).nodeLink; 
    } 
   
    //print stack contents
    public void display() { 
        // check for stack underflow 
        if (topNode == null) { 
            System.out.printf("\nStack Underflow!!"); 
            exit(1); 
        } 
        // if the top node is not null
        else { 
            // set the top node as the temporary node
            Node tempNode = topNode;
            System.out.println("Stack elements:");
            // if the temporary node is not null
            while (tempNode != null) { 
                  // print value of node
                System.out.print(tempNode.value + " -> ");
                // assign temporary link to temporary node
                tempNode = tempNode.nodeLink; 
            } 
        } 
    } 
} 
 
public class Main { 
    public static void main(String[] args) { 
        // Create a stack class object 
        stackLinkedlist stackLinkedListObject = new stackLinkedlist(); 

        Scanner userInput = new Scanner(System.in);

        System.out.println("Enter the amount of elements to add to Stack: ");
        int numElements = userInput.nextInt();
        
        while(numElements != 0) {
            numElements  = numElements  - 1;
            System.out.println("Enter element to input into stack: ");
            int element = userInput.nextInt();
            stackLinkedListObject.push(element);
        }

        // print Stack elements 
        stackLinkedListObject.display(); 
   
        // print current element at the top of the Stack
        System.out.println("\nStack top: " + stackLinkedListObject.peek());
 
        // Pop elements twice 
        System.out.println("Pop two elements");
        stackLinkedListObject.pop(); 
        stackLinkedListObject.pop();
   
        // print elements in the Stack
        stackLinkedListObject.display(); 
   
        // print new element at top of Stack
        System.out.println("\nNew Stack top: " + stackLinkedListObject.peek()); 
    } 
}



/*
    Example Input

    Enter the amount of elements to add to Stack: 5

    Enter element to input into stack: 
    9
    7
    5
    3
    1


    Expected Output

    Stack elements:
    1 -> 3 -> 5 -> 7 -> 9 -> 
    Stack top: 1
    Pop two elements
    Stack elements:
    5 -> 7 -> 9 -> 
    New Stack top: 5
 */

File: LInear Data Structures/StackLinkedList.cpp
// stack using singly linked list

#include <bits/stdc++.h>
using namespace std;

// creating a linked list;
class Node {
public:
	int data;
	Node* link;

	// Constructor
	Node(int n)
	{
		this->data = n;
		this->link = NULL;
	}
};

class Stack {
	Node* top;

public:
	Stack() { top = NULL; }

	void push(int data)
	{

		// Create new node temp and allocate memory in heap
		Node* temp = new Node(data);

		// Check if stack (heap) is full.
		// Then inserting an element would
		// lead to stack overflow
		if (!temp) {
			cout << "\nStack Overflow";
			exit();
		}

		// Initialize data into temp data field
		temp->data = data;

		// Put top pointer reference into temp link
		temp->link = top;

		// Make temp as top of Stack
		top = temp;
	}

	// check if Stack is empty
	bool isEmpty()
	{
		// If top is NULL it means that
		// there are no elements are in stack
		return top == NULL;
	}

	// remove element from Stack
	void pop()
	{
		Node* temp;

		// Check for stack underflow
		if (top == NULL) {
			cout << "\nStack Underflow" << endl;
			exit();
		}
		else {

			// Assign top to temp
			temp = top;

			// Assign second node to top
			top = top->link;

			// This will automatically destroy
			// the link between first node and second node

			// Release memory of top node
			// i.e delete the node
			free(temp);
		}
	}

	// print all elements in Stack
	void display()
	{
		Node* temp;

		// Check for stack underflow
		if (top == NULL) {
			cout << "\nStack Underflow";
		}
		else {
			temp = top;
			while (temp != NULL) {

				// Print node data
				cout << temp->data;

				// Assign temp link to temp
				temp = temp->link;
				if (temp != NULL)
					cout << " -> ";
			}
		}
	}

	    void exit() {
        cout << "Hope to see you again soon!" << endl;
    }
};

// main driver method
int main()
{
	Stack userStack;

	int choice;
	int value;
	
	do {
		cout << "\nSelect an option:\n1. Enter element to Stack \n2. Pop the element from Stack \n3. Display all the elements in the Stack \n0. Exit\n";
		cin >> choice;

		switch (choice) {
			case 0:
				userStack.exit();
				break;
			case 1:
				cout << "Enter the number you want to enter into Stack ";
				cin >> value;
				userStack.push(value);
				break;
				
			case 2:
				userStack.pop();
				break;

			case 3:
				userStack.display();
                break;

			default:
				cout << "Please make another choice\n";
				break;
			}
	} 
	
	while (choice != 0);
		return 0;
}
